consumer/file.go

```diff
@@ -186,85 +180,98 @@ func (cons *File) initFile() {
+		cons.watcher = newWatcher(cons.Logger, &cons.source, func() { cons.read(buffer, sendFunction, func() {}, func() {}) })
+		cons.watcher.Watch(buffer, sendFunction)
+	} else {
+		cons.poll(buffer, sendFunction)
+	}
+}
 
-		// Initialize the seek state if requested
-		// Try to read the remains of the file first
-		if cons.state == fileStateOpen {
-			if cons.file != nil {
-				buffer.ReadAll(cons.file, sendFunction)
-			}
-			cons.initFile()
-			buffer.Reset(uint64(cons.seekOffset))
-		}
+func (cons *File) poll(buffer *tio.BufferedReader, sendFunction func(data []byte)) {
+	spin := tsync.NewCustomSpinner(cons.source.pollingDelay)
+
+	for cons.source.state != fileStateDone {
+		cons.read(buffer, sendFunction, spin.Yield, spin.Reset)
+	}
+}
 
-		// Try to open the file to read from
-		if cons.state == fileStateRead && cons.file == nil {
-			file, err := os.OpenFile(cons.realFileName, os.O_RDONLY, 0666)
+func (cons *File) read(buffer *tio.BufferedReader, sendFunction func(data []byte), onEOF func(), onAfterRead func()) {
+	// Initialize the seek state if requested
+	// Try to read the remains of the file first
+	if cons.source.state == fileStateOpen {
+		if cons.source.file != nil {
+			buffer.ReadAll(cons.source.file, sendFunction)
+		}
+		cons.initFile()
+		buffer.Reset(uint64(cons.seeker.offset))
+	}
 
-			switch {
-			case err != nil:
-				if printFileOpenError {
-					cons.Logger.Warning("Open failed: ", err)
-					printFileOpenError = false
-				}
-				time.Sleep(3 * time.Second)
-				continue // ### continue, retry ###
+	// Try to open the file to read from
+	if cons.source.state == fileStateRead && cons.source.file == nil {
+		file, err := os.OpenFile(cons.source.realFileName, os.O_RDONLY, 0666)
 
-			default:
-				cons.file = file
-				cons.seekOffset, _ = cons.file.Seek(cons.seekOffset, cons.seek)
-				printFileOpenError = true
+		switch {
+		case err != nil:
+			if cons.source.printFileOpenError {
+				cons.Logger.Warning("Open failed: ", err)
+				cons.source.printFileOpenError = false
 			}
+			time.Sleep(3 * time.Second)
+			return // ### continue, retry ###
+
+		default:
+			cons.source.file = file
+			cons.seeker.offset, _ = cons.source.file.Seek(cons.seeker.offset, cons.seeker.seek)
+			cons.source.printFileOpenError = true
 		}
+	}
 
-		// Try to read from the file
-		if cons.state == fileStateRead && cons.file != nil {
-			err := buffer.ReadAll(cons.file, sendFunction)
+	// Try to read from the file
+	if cons.source.state == fileStateRead && cons.source.file != nil {
+		err := buffer.ReadAll(cons.source.file, sendFunction)
 
-			switch {
-			case err == nil: // ok
-				spin.Reset()
+		switch {
+		case err == nil: // ok
+			onAfterRead()
 
-			case err == io.EOF:
-				if cons.file.Name() != cons.realFileName {
+		case err == io.EOF:
+			if cons.source.file.Name() != cons.source.realFileName {
+				cons.Logger.Info("Rotation detected")
+				cons.onRoll()
+			} else {
+				newStat, newStatErr := os.Stat(cons.source.realFileName)
```

Comment:
This will still call os.Stat every loop. To avoid this, introduce a timer or counter and only call os.Stat every x.
e.g.
```go
lastStat := time.Now()
// ...
if time.Since(lastStat) > time.Second {
  // ... do os.Stat stuff
  lastStat = time.Now()
}
```